[ntuple] fix type name normalization (#17637)
root-project/root: tree/ntuple/v7/doc/BinaryFormatSpecification.md

Within the supported type system complex types can be freely composed,
e.g. `std::vector<MyEvent>` or `std::vector<std::vector<float>>`.

### Type Name Normalization

Type names are stored according to the following normalization rules

>   - All whitespaces are removed.

Reviewer: Not exact phrasing, in part because of https://github.com/root-project/root/pull/17637/files#r1943506781

Author: Clarified.

>   - CV qualifiers (`const`, `volatile`) are removed.

Reviewer: Only the outer ones can be safely removed: `usertemplate<const int>` might have a complete different layout than `usertemplate<int>`

Author: Thanks, good point! Fixed.

>   - All `volatile` qualifiers are removed; `const` qualifiers that do not appear in template arguments are removed.

Reviewer: According to this [example](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:13,endLineNumber:14,positionColumn:13,positionLineNumber:14,selectionStartColumn:13,selectionStartLineNumber:14,startColumn:13,startLineNumber:14),source:'%23include+%3Coptional%3E%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+Device+%7B%0A++++int+stack_size+%3D+0%3B%0A++++int+heap_size+%3D+0%3B%0A%7D%3B%0A%0Aint+main()+%7B%0A++++Device%3Cint%3E+d0%3B%0A++++Device%3Cvolatile+int%3E+d1%3B%0A++++Device%3Cconst+int%3E+d2%3B%0A++++//+d1+%3D+d2%3B%0A++++d0+%3D+d1%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:51.145194274028626,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g132,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17+-Wall+-Wextra+-Wpedantic',overrides:!(),runtimeTools:!(),source:1,stdinPanelShown:'1',tree:0,wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+13.2+(C%2B%2B,+Editor+%231)',t:'0')),header:(),k:48.85480572597136,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4). 
the types are also different for volatile being part of the template arguments.

>   - All whitespaces except between `const` and a template argument are removed.

Reviewer: ```suggestion
  - All whitespaces not required by C++ semantic are removed.
```

  - The `class`, `struct`, and `enum` keywords are removed.
  - Type names are fully qualified by the namespace in which they are declared.

>   - The integer types `signed char`, `unsigned char`, and `[signed|unsigned](short|int|long[ long])` are replaced by

Reviewer: What about `cont char*`?

Author: `char *` and by extension `const char *` are not supported by RNTuple. Users need to use `std::string`.

    the corresponding (at the time of writing) `std::[u]int(8|16|32|64)_t` standard integer typedef.
  - Type names used as template arguments are themselves normalized.
  - For user-defined types, default-initialized template arguments are explicitly spelled out;

>     optional template arguments of stdlib types such as allocators are omitted.

Reviewer: Note that the I/O normalization we have elected to keep the stdlib optional argument whose value is not the default value.  ie.:
```
TClass::GetClass("vector<int, other_alloc>")->GetName()
(const char *) "vector<int,other_alloc>"
```

Author: Ok, I think that should be solved when fixing #16754

  - Integer template arguments are written in standard decimal notation; only if necessary (long integer types),
    a suffix `l` or `ul` is added according to the template parameter type.

> The type name of a field has typedefs and usings fully resolved (except for standard integer typedefs).

Reviewer: This may belong sooner in the document as the above mentioned rules are applying to the name with fully resolved typedefs.

Author: Hm, I'm not sure I understand. The normalization rules, as far as I understand, should apply to both the original type name and the resolved type name.

Reviewer: The replacement of the typedef is an integral type of the 'normalization' of the name.  Without that part, the name are not unique and thus not normalized.  I.e. if you have both `const vector<vector<Int_t> >` and `vector<vector<int32_t, typedef_to_regular_allocator>>` "only" pretty printing the name does not help much.  Similar if the typedef are not replaced, neither would/should the Integer template arguments be calculated/compacted.  In the second example, without resolving the typedef we can not even remove the default template argument.

Author: If the user passes us `vector<vector<int32_t, typedef_to_regular_allocator>>`, its "pretty printed" version will be registered as the field's alias type, i.e. the "original spelling" from the user. Otherwise (`vector<vector<int>>`), the alias type is empty.

Reviewer: For the original spelling, I guess we ought to decide to either really put it as-is or we need to spell out, separately from the normalization, what applies there.
For example, it is clear that the 'whitespace' rule would apply,  maybe the `const` rule and what else?  (Not the `typedef` for sure, what about the `std::[u]int(8|16|32|64)_t` or even the fully qualified)

>   - All `volatile` qualifiers are removed; `const` qualifiers that do not appear in template arguments are removed.
>   - All whitespaces except between `const` and a template argument are removed.
>   - The `class`, `struct`, and `enum` keywords are removed.
>   - Type names are fully qualified by the namespace in which they are declared.
>   - The integer types `signed char`, `unsigned char`, and `[signed|unsigned](short|int|long[ long])` are replaced by
>     the corresponding (at the time of writing) `std::[u]int(8|16|32|64)_t` standard integer typedef.
>   - Type names used as template arguments are themselves normalized.
>   - For user-defined types, default-initialized template arguments are explicitly spelled out;
>     optional template arguments of stdlib types such as allocators are omitted.
>   - Integer template arguments are written in standard decimal notation; only if necessary (long integer types),
>     a suffix `l` or `ul` is added according to the template parameter type.
> 
> The type name of a field has typedefs and usings fully resolved (except for standard integer typedefs).

Reviewer: ```suggestion
  - The type name of a field has typedefs and usings fully resolved (except for the following rule).
  - The integer types `signed char`, `unsigned char`, and `[signed|unsigned](short|int|long[ long])` are replaced by
    the corresponding (at the time of writing) `std::[u]int(8|16|32|64)_t` standard integer typedef.
  -`volatile` and `const` qualifiers that do not appear in template arguments are removed.
  - The `class`, `struct`, and `enum` keywords are removed.
  - Type names are fully qualified by the namespace in which they are declared.
  - Type names used as template arguments are themselves normalized.
  - For user-defined types, default-initialized template arguments are explicitly spelled out;
    optional template arguments of stdlib types such as allocators are omitted.
  - Integer template arguments are written in standard decimal notation; only if necessary (long integer types),
    a suffix `l` or `ul` is added according to the template parameter type.
  - Only whitespaces strictly required by C++ semantic are kept.
```
